tools/flakiness_checker.py

```diff
@@ -0,0 +1,87 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+""" Checks a given test for flakiness
+Takes the file name and function name of a test, as well as, optionally,
+the number of trials to run and the random seed to use
+"""
+
+import subprocess
+import sys
+import os
+import random
+import argparse
+
+
+DEFAULT_NUM_TRIALS = 500
```

Comment:
Flakyness Test should be at least 10000 runs

Reply:
0.99^500 < 0.01, which means that a flaky test that succeeds with 99% of probability has less than 1% chance of succeeding 500 runs.

Reply:
BTW this is just the default value, for this tool you can always provide an argument to specify how many trials you want.

Reply:
Actually some unit tests may take up to more than 30 seconds each, if you're running it for more than 10000 times by default then that's 300000 seconds which is ~3.5days. Say if some tests is related to random seeds, it's due to input values generated randomly by the seed which is also randomly drawn, that's like a fixed probability. BTW 0.999^7000 < 0.001 so we don't even need 10000 times I guess...

Reply:
If a test takes 30 seconds, it's not a unit test but rather an integration test and should then not rely on randon data. 

The majority of our tests run in less than a second and we should design the default values for these cases. We employ the same concept for most of our automation scripts - for example, it's sufficient to run CI/build.py -p armv7 and that's it. The default behaviour should be the main use cases and everything specialized (like a long running test) should require an additional Parameter. In general, it's best practice to keep the number of configuration options for the standard case to an absolute minimum.

Reply:
Our software runs in services that that serve millions of requests - even a 0.1% chance can be devastating. We should err on the side of caution as a default.

Reply:
On the other hand, I think the main purpose of running tests is to check your own implementation rather than making tests pass all the time. For discovering problems in your own code, less number of trials is sufficient. In addition, when we deploy this on CI we also don't want it to introduce much overhead in the CI. As it is stated in the design doc of Carl's project, the golden standard of checking flakiness is running it infinite number of times or at least with all possible seeds (0-2147483647), but it's also something impossible.

Reply:
Well, there're a lot of unit tests that would last for more than 30 seconds...even it only lasts for 3.5 seconds each, 10000 runs would cost ~10hrs.

Reply:
Well the purpose of this script is specifically to detect flaky tests, right? I'd expect that the expectation would then be that it passes all the time.

Sure, we won't achieve perfect coverage and at some points you get diminishing returns, but I have seen that even you usually use 10000 runs as the standard value for your tests. Time should also not be that much of a factor (even a few hours are fine) because this is intended to he run in the background. Robustness tests are a time intensive task and they should be designed to leave as few room for errors as possible (with costs in mind).

For CI you don't have to worry, this will run asynchronously and we'll make a proper assessment what are be best values for which case - it will probably be a Time-Boxed execution. But we are not at that stage yet.

Fyi, we got 455 tests < 0.1s, 509 < 1s, 565 < 3s and 63 > 3s. 628 tests in total.

Reply:
But is this supposed to be a required check for every PR? If it's taking too long then is it becoming an obstacle for PRs.